File: tools/commonist/src/scala/commonist/util/Loader.scala
package commonist.util

import java.io.File
import java.io.IOException
import java.net.MalformedURLException
import java.net.URL

import scutil.ext.AnyRefExt._
import scutil.ext.OptionExt._

/** loads resources from a set of URL-paths */
final class Loader(settingsDir:File, projectDir:File, resourcePrefix:String) {
	// TODO re-introduce debug output with noneEffect and someEffect in OptionExt
	def resourceURL(path:String):Option[URL] = 
			directoryURL(settingsDir, path) orElse 
			directoryURL(projectDir, path) orElse
			classloaderURL(resourcePrefix, path)
	
	def directoryURL(directory:File, path:String):Option[URL] =
			new File(directory, path) match {
				case file if file.exists	=> Some(file.toURI.toURL)
				case _						=> None
			}
			
	def classloaderURL(resourcePrefix:String, path:String):Option[URL] =
			getClass getResource (resourcePrefix + path) nullOption
}

File: tools/commonist-java/src/net/psammead/commonist/util/Loader.java
package net.psammead.commonist.util;

import java.io.File;
import java.io.IOException;
import java.net.MalformedURLException;
import java.net.URL;

import net.psammead.util.Logger;

/** loads resources from a set of URL-paths */
public final class Loader {
	private static final Logger log = new Logger(Loader.class);

	private final File	settingsDir;
	private final File	projectDir;
	private final String	resourcePrefix;

	public Loader(File settingsDir, File projectDir, String resourcePrefix) {
		this.settingsDir	= settingsDir;
		this.projectDir		= projectDir;
		this.resourcePrefix	= resourcePrefix;
	}
	
	/** returns a resourceURL or throws an IOException */
	public URL mandatoryURL(String path) throws IOException {
		final URL	url	= optionalURL(path);
		if (url == null)	throw new IOException("cannot locate: " + path);
		return url;
	}
	
	/** returns null when a resource does not exist */
	public URL optionalURL(String path) {
		try {
			final File	settingsFile	= new File(settingsDir, path);
			if (settingsFile.exists()) {
				log.debug("loading from settings: " + path);
				return settingsFile.toURI().toURL();
			}
		}
		catch (MalformedURLException e) {
			log.error("cannot examine", e);
		}
		
		try {
			final File	projectFile	= new File(projectDir, path);
			if (projectFile.exists()) {
				log.debug("loading from project: " + path);
				return projectFile.toURI().toURL();
			}
		}
		catch (MalformedURLException e) {
			log.error("cannot examine", e);
		}
		
		final URL	url	= getClass().getResource(resourcePrefix + path);
		if (url != null) {
			log.debug("loading from classloader: " + path);
			return url;
		}
		
		return null;
	}
}
